cs3api4lab/api/cs3_file_api.py

```diff
@@ -261,6 +261,64 @@ def create_directory(self, path, endpoint=None):
         self.log.debug(
             'msg="Invoked create container" filepath="%s" elapsedTimems="%.1f"' % (path, (tend - tstart) * 1000))
 
+    def list_checkpoints(self, path):
+        print('list path', path)
+        ref = FileUtils.get_reference(path)
+        request =cs3sp.ListFileVersionsRequest(ref=ref)
+        response = self.cs3_api.ListFileVersions(request=request, metadata=[('x-access-token', self.auth.authenticate())])
+
+        print('list response', response)
+        if response.status.code != cs3code.CODE_OK:
+            self.log.error(f"msg=\"Failed to list file versions\" filepath={path}")
+
+        checkpoints = []
+        for i in reversed(range(len(response.versions)-1)):
+            if self.config.dev_env: # we need this until https://github.com/cs3org/reva/issues/3928 is resolved
+                last_modified = datetime.datetime.fromtimestamp(response.versions[i].mtime / 1000.0)
+            else:
+                last_modified = datetime.datetime.fromtimestamp(response.versions[i].mtime)
+
+            # because the checkpoint url is parsed by regular expression, so we need to replace the '.' sign from key
+            key = str(response.versions[i].key).replace('.', '-')
+
+            checkpoints.append({
+                'id': key,
+                'last_modified': last_modified
+            })
+
+        return checkpoints
+
+    def restore_checkpoint(self, checkpoint_id, path):
+        # because the checkpoint url is parsed by regular expression, so we need to replace the '.' sign from key
+        checkpoint_id = str(checkpoint_id).replace('-', '.')
+        if self.config.dev_env: # we need this until https://github.com/cs3org/reva/issues/3927 is fixed
+            checkpoint_id = 'v' + checkpoint_id
+
+        ref = FileUtils.get_reference(path, self.config.endpoint)
+        request = cs3sp.RestoreFileVersionRequest(ref=ref, key=checkpoint_id)
```

Comment:
Can we move this request bellow (inside the if == metadata)? I think it becomes a bit clearer.

Reply:
The request object is used for metadata and for cs3 locks as well (for a file without a lock). I've moved it a little lower, just before if statement so it would be more readable.